### ISSUE TITLE
Should MatMul op allow specification of accumulation precision?

### ISSUE THREAD

**Issue by gramalingam** [onnx/onnx#7072]
The use of lower-precision floating points is becoming more common. Use of 16-bit float matmul is common. However, in practice, use of 32-bit accumulation is very useful for 16-bit float matmul. The ONNX matmul op spec does not specify anything about the accumulation precision. 

It would be useful to augment the op with a new optional attribute to specify the accumulation precision. If no attribute is specified, accumulation precision would be the input precision (the current interpretation, made explicit).

Same thing would apply to Gemm.

Similar issues show up in other reduction ops, like Softmax, as well.

**Comment by justinchuby** [comment]
@onnx/sig-operators

**Comment by computer-whisperer** [comment]
I ran into this recently with bf16 tensors in my own onnx runtime. Pytorch automatically upcasts to fp32 on a number of ops, and any models trained to that standard will develop accuracy issues if that doesn't happen. Making this decision explicit rather than relying on the runtime to automatically upcast on certain types would be much preferred.
